File: editor/src/main/kotlin/de/fatox/meta/ide/AssetDiscoverer.java
package de.fatox.meta.ide;

import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import de.fatox.meta.Meta;
import de.fatox.meta.api.ui.UIManager;
import de.fatox.meta.injection.Inject;
import de.fatox.meta.ui.windows.AssetDiscovererWindow;

/**
 * Created by Frotty on 07.06.2016.
 */
public class AssetDiscoverer {
    @Inject
    private UIManager uiManager;
    @Inject
    private ProjectManager projectManager;

    private FileHandle root;

    private FileHandle currentFolder;

    private Array<FileHandle> currentChildFolders;

    private Array<FileHandle> currentChildFiles;

    private ObjectMap<String, AssetOpenListener> fileOpenListeners = new ObjectMap<>();

    public AssetDiscoverer() {
        Meta.inject(this);
    }

    public void openFolder(FileHandle fileHandle) {
        currentFolder = fileHandle;
        refresh();
    }

    public void openChild(String name) {
        currentFolder = currentFolder.child(name);
        refresh();
    }

    public Array<FileHandle> getCurrentChildFolders() {
        return currentChildFolders;
    }

    public Array<FileHandle> getCurrentChildFiles() {
        return currentChildFiles;
    }

    public void refresh() {
        currentChildFolders = new Array<>();
        currentChildFiles = new Array<>();
        for (FileHandle child : currentFolder.list()) {
            if (child.isDirectory()) {
                currentChildFolders.add(child);
            } else {
                currentChildFiles.add(child);
            }
        }
        AssetDiscovererWindow window = uiManager.getWindow(AssetDiscovererWindow.class);
        if (window != null) {
            window.refresh();
        }
    }

    public void openFile(FileHandle fileHandle) {
        if (fileOpenListeners.containsKey(fileHandle.extension())) {
            fileOpenListeners.get(fileHandle.extension()).onOpen(fileHandle);
        }
    }

    public FileHandle getCurrentFolder() {
        return currentFolder;
    }

    public FileHandle getRoot() {
        return root;
    }

    public void addOpenListener(String extension, AssetOpenListener listener) {
        fileOpenListeners.put(extension, listener);
    }

    public void setRoot(String path) {
        if (path.length() <= 0) {
            currentFolder = projectManager.getCurrentProjectRoot();
        } else {
            currentFolder = projectManager.getCurrentProjectRoot().child(path);
        }
        this.root = currentFolder;
        refresh();
    }

    public void setRoot(FileHandle root) {
        this.root = root;
        currentFolder = root;
        refresh();
    }
}

File: editor/src/de/fatox/meta/ide/AssetDiscoverer.kt
package de.fatox.meta.ide

import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ObjectMap
import de.fatox.meta.api.ui.UIManager
import de.fatox.meta.injection.MetaInject.Companion.lazyInject
import de.fatox.meta.ui.windows.AssetDiscovererWindow

/**
 * Created by Frotty on 07.06.2016.
 */
class AssetDiscoverer {
	private val uiManager: UIManager by lazyInject()
	private val projectManager: ProjectManager by lazyInject()

	var root: FileHandle? = null
		private set
	var currentFolder: FileHandle? = null
		private set
	var currentChildFolders: Array<FileHandle>? = null
		private set
	var currentChildFiles: Array<FileHandle>? = null
		private set
	private val fileOpenListeners = ObjectMap<String, AssetOpenListener>()
	fun openFolder(fileHandle: FileHandle?) {
		currentFolder = fileHandle
		refresh()
	}

	fun openChild(name: String?) {
		currentFolder = currentFolder!!.child(name)
		refresh()
	}

	fun refresh() {
		currentChildFolders = Array()
		currentChildFiles = Array()
		for (child in currentFolder!!.list()) {
			if (child.isDirectory) {
				currentChildFolders!!.add(child)
			} else {
				currentChildFiles!!.add(child)
			}
		}
		val window = uiManager.getWindow(AssetDiscovererWindow::class.java)
		window.refresh()
	}

	fun openFile(fileHandle: FileHandle) {
		if (fileOpenListeners.containsKey(fileHandle.extension())) {
			fileOpenListeners.get(fileHandle.extension()).onOpen(fileHandle)
		}
	}

	fun addOpenListener(extension: String, listener: AssetOpenListener) {
		fileOpenListeners.put(extension, listener)
	}

	fun setRoot(path: String) {
		currentFolder = if (path.length <= 0) {
			projectManager.currentProjectRoot
		} else {
			projectManager.currentProjectRoot.child(path)
		}
		this.root = currentFolder
		refresh()
	}

	fun setRoot(root: FileHandle?) {
		this.root = root
		currentFolder = root
		refresh()
	}
}